src/client/cl_tempentities.c

```diff
@@ -1377,318 +1419,207 @@ CL_AddBeams(void)
 
 			if (b->model == cl_mod_lightning)
 			{
-				ent.flags = RF_FULLBRIGHT;
 				ent.angles[0] = -pitch;
 				ent.angles[1] = yaw + 180.0f;
-				ent.angles[2] = (float)(randk() % 360);
 			}
 			else
 			{
 				ent.angles[0] = pitch;
 				ent.angles[1] = yaw;
-				ent.angles[2] = (float)(randk() % 360);
 			}
 
-			V_AddEntity(&ent);
+			ent.angles[2] = (float)(randk() % 360);
 
-			for (j = 0; j < 3; j++)
-			{
-				org[j] += dist[j] * len;
-			}
+			V_AddEntity(&ent);
 
+			VectorMA(org, len, dist, org);
 			d -= model_length;
 		}
 	}
 }
 
-extern cvar_t *hand;
+static int
+HandPos(void)
+{
+	int v;
 
-void
-CL_AddPlayerBeams(void)
+	if (!hand)
+	{
+		return 0;
+	}
+
+	v = hand->value;
+
+	return (v <= 2 && v >= 0) ? v : 0;
+}
+
+static void
+ApplyBeamOffset(const vec3_t ofs, float hand_mul, vec3_t out)
 {
-	int i, j;
-	beam_t *b;
+	VectorMA(out, hand_mul * ofs[0], cl.v_right, out);
+	VectorMA(out, ofs[1], cl.v_forward, out);
+	VectorMA(out, ofs[2], cl.v_up, out);
+
+	if (hand_mul == 0.0f)
+	{
+		VectorMA(out, -1, cl.v_up, out);
+	}
+}
+
+static void
+AdjustToWeapon(const vec3_t ofs, float hand_mul, vec3_t out)
+{
+	int i;
+	frame_t *oldframe;
+	player_state_t *ps, *ops;
+
+	/* set up gun position */
+	ps = &cl.frame.playerstate;
+	i = (cl.frame.serverframe - 1) & UPDATE_MASK;
+	oldframe = &cl.frames[i];
+
+	if ((oldframe->serverframe != cl.frame.serverframe - 1) || !oldframe->valid)
+	{
+		oldframe = &cl.frame; /* previous frame was dropped or invalid */
+	}
+
+	ops = &oldframe->playerstate;
+
+	for (i = 0; i < 3; i++)
+	{
+		out[i] = cl.refdef.vieworg[i] + ops->gunoffset[i]
+				  + cl.lerpfrac * (ps->gunoffset[i] - ops->gunoffset[i]);
+	}
+
+	ApplyBeamOffset(ofs, hand_mul, out);
+}
+
+static void
+CL_AddHeatBeams(void)
+{
+	const beam_t *b;
 	vec3_t dist, org;
 	float d;
 	entity_t ent;
 	float yaw, pitch;
-	float forward;
 	float len, steps;
 	int framenum;
 	float model_length;
-
-	float hand_multiplier;
-	frame_t *oldframe;
-	player_state_t *ps, *ops;
+	int by_us;
+	int hand_pos;
+	float hand_mul;
 
 	framenum = 0;
 
-	if (hand)
-	{
-		if (hand->value == 2)
-		{
-			hand_multiplier = 0;
-		}
+	hand_pos = HandPos();
 
-		else if (hand->value == 1)
-		{
-			hand_multiplier = -1;
-		}
-
-		else
-		{
-			hand_multiplier = 1;
-		}
-	}
-	else
+	switch (hand_pos)
```

Comment:
I'm too old and dumb to understand this code, but since this is the only usage of `HandPos()`, just to set `hand_mul`, shouldn't `HandPos()` itselt return the correct value for `hand_mul` directly? The `switch()` feels unnecessary, and you can even return an `int` if the data type is important.

In other topic, with `hand 1` (left), the plasma beam shoots wherever, ahead of where you are pointing with the mouse; something like a Wii shooter. `0` and `2` work just fine.

Reply:
Yeah, it looks pretty good now, with no apparent "misbehavior". Thanks for your work!